### PR TITLE
Translate & populate org_id when missing from host payloads

### PR DESCRIPTION
**Pull request by kruai**
# Overview

This PR is being created to address [ESSNTL-2775](https://issues.redhat.com/browse/ESSNTL-2775).
Makes it so that if an incoming host is missing `org_id`, it attempts to fetch it from the tenant-translation 3scale endpoint. It uses the same retry & timeout config that the RBAC requests use. If the org_id is missing from the host and it's unable to fetch the value from the tenant translator, the payload is rejected.

On hold until #1125 is merged.

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -15,7 +15,7 @@ run_inv_web_service:
 	#
 	# KAFKA_TOPIC="platform.system-profile" KAFKA_GROUP="inventory" KAFKA_BOOTSTRAP_SERVERS="localhost:29092"
 	#
-	INVENTORY_LOG_LEVEL=DEBUG BYPASS_RBAC=true gunicorn -b :8080 run
+	INVENTORY_LOG_LEVEL=DEBUG BYPASS_RBAC=true BYPASS_TENANT_TRANSLATION=true gunicorn -b :8080 run
```
Comment (thearifismail) [inline Makefile:18]: `BYPASS_TENANT_TRANSLATION` could be replaced by `SET_ORG_ID` for clarity and avoid the use double negatives

File: app/config.py [hunk 2]
```diff
@@ -107,6 +107,9 @@ def __init__(self, runtime_environment):
         self.rbac_retries = os.environ.get("RBAC_RETRIES", 2)
         self.rbac_timeout = os.environ.get("RBAC_TIMEOUT", 10)
 
+        self.tenant_translator_url = os.environ.get("TENANT_TRANSLATOR_URL")
+        self.bypass_tenant_translation = os.environ.get("BYPASS_TENANT_TRANSLATION", "false").lower() == "true"
```
Comment (thearifismail) [inline app/config.py:111]: Using `bypass=false` mean org_id would be set by default?  Is it what the intent is and when will it go in effect?

**Comment by fstavela** [comment]
@kruai Is there a documentation for the tenant translator somewhere?

File: app/instrumentation.py [hunk 3]
```diff
@@ -186,6 +187,11 @@ def rbac_permission_denied(logger, required_permission, user_permissions):
     rbac_access_denied.labels(required_permission=required_permission).inc()
 
 
+def tenant_translator_failure(logger: str, error_message: str = None):
```
Comment (fstavela) [inline app/instrumentation.py:190]: You probably didn't mean to set logger type to str? It should be Logger from logging.

**Review by fstavela — COMMENTED** [review]
PR check is failing because we don't deploy 3scale in ephemeral envs and HBI tries to make requests in every attempt to create/update a host. I'm trying to deploy and test it locally, but I'm having troubles. I see this in HBI logs when attempting to create a host:
```
[2022-06-02 11:53:46,078] [1] [140002213982464] [inventory.app.queue.queue] [INFO] Attempting to add host
[2022-06-02 11:53:46,078] [1] [140002213982464] [inventory.app.queue.queue] [ERROR] Error while adding host
Traceback (most recent call last):
  File "/opt/app-root/src/app/queue/queue.py", line 235, in add_host
    output_host, host_id, insights_id, add_result = host_repository.add_host(
  File "/opt/app-root/src/lib/host_repository.py", line 61, in add_host
    input_host.org_id = translate_account_to_org_id(input_host.account)
  File "/opt/app-root/src/lib/middleware.py", line 113, in translate_account_to_org_id
    request_session.mount(tenant_translator_url(), HTTPAdapter(max_retries=retry_config))
  File "/opt/app-root/lib64/python3.8/site-packages/requests/sessions.py", line 755, in mount
    keys_to_move = [k for k in self.adapters if len(k) < len(prefix)]
  File "/opt/app-root/lib64/python3.8/site-packages/requests/sessions.py", line 755, in <listcomp>
    keys_to_move = [k for k in self.adapters if len(k) < len(prefix)]
TypeError: object of type 'NoneType' has no len()
[2022-06-02 11:53:46,080] [1] [140002213982464] [inventory.app.queue.queue] [ERROR] Unable to process message
Traceback (most recent call last):
  File "/opt/app-root/src/app/queue/queue.py", line 298, in event_loop
    handler(message.value, event_producer)
  File "<decorator-gen-6>", line 2, in handle_message
  File "/opt/app-root/lib64/python3.8/site-packages/prometheus_client/context_managers.py", line 66, in wrapped
    return func(*args, **kwargs)
  File "/opt/app-root/src/app/queue/queue.py", line 272, in handle_message
    output_host, host_id, insights_id, operation_result = message_operation(host, platform_metadata)
  File "/opt/app-root/src/app/queue/queue.py", line 235, in add_host
    output_host, host_id, insights_id, add_result = host_repository.add_host(
  File "/opt/app-root/src/lib/host_repository.py", line 61, in add_host
    input_host.org_id = translate_account_to_org_id(input_host.account)
  File "/opt/app-root/src/lib/middleware.py", line 113, in translate_account_to_org_id
    request_session.mount(tenant_translator_url(), HTTPAdapter(max_retries=retry_config))
  File "/opt/app-root/lib64/python3.8/site-packages/requests/sessions.py", line 755, in mount
    keys_to_move = [k for k in self.adapters if len(k) < len(prefix)]
  File "/opt/app-root/lib64/python3.8/site-packages/requests/sessions.py", line 755, in <listcomp>
    keys_to_move = [k for k in self.adapters if len(k) < len(prefix)]
TypeError: object of type 'NoneType' has no len()
```
I'm not yet sure if this is because of a wrong 3scale deployment, or if this is a bug in HBI. It looks like this is going to be an "interesting" time spent on debugging environment issues. I hope that 3scale is properly clowderized.

File: app/instrumentation.py (hunk 3 above)
Comment (thearifismail) [reply app/instrumentation.py:190]: `str` is just a non-ideal variable name; its type is `Logger`

File: tests/fixtures/api_fixtures.py [hunk 4]
```diff
@@ -83,6 +83,11 @@ def enable_rbac(inventory_config):
     inventory_config.bypass_rbac = False
 
 
+@pytest.fixture(scope="function")
+def enable_org_id_translation(inventory_config):
+    inventory_config.bypass_tenant_translation = False
```
Comment (thearifismail) [inline tests/fixtures/api_fixtures.py:88]: This sets the default behavior to "set org_id" but when `test_add_host_missing_org_id` is executed and `def translate_account_to_org_id()` is called, the `tenant_translator_url()` returns `None`

File: app/instrumentation.py (hunk 3 above)
Comment (fstavela) [reply app/instrumentation.py:190]: Yeah, caller provides `Logger`, but the function expects `str`. I guess it's not going to result in an error in Python, but some type checker would probably complain.

File: lib/middleware.py [hunk 5]
```diff
@@ -94,3 +101,32 @@ def modified_func(*args, **kwargs):
...
+        return None
+
+    request_session = Session()
+    retry_config = Retry(total=inventory_config().rbac_retries, backoff_factor=1, status_forcelist=RETRY_STATUSES)
+    request_session.mount(tenant_translator_url(), HTTPAdapter(max_retries=retry_config))
+    body = dumps({"body": "[{" + account + "}]"})
+
+    try:
+        with outbound_http_metric.time():
+            translator_response = request_session.post(
```
Comment (thearifismail) [inline lib/middleware.py:118]: What's the purpose of this `post` to `tenant_url` which I think post to the 3Scale service.  Yesterday I thought I saw a reference to 3Scale for get an ORG_ID for the account but now I don't see that reference.  Based on the PR description `3Scale` endpoint is used but if an `org_id` not found, does it mean that then the value of `resp_data[account]` will be `None`?

**Review by thearifismail — COMMENTED** [review]
Please see comments

File: Makefile (hunk 1 above)
Comment (kruai) [reply Makefile:18]: I could change it to `ENABLE_TENANT_TRANSLATION`, but I wanted to imply that the default functionality is to translate account numbers, and that we specifically have to set a flag to disable it (same reasoning as why we started using `BYPASS_RBAC`). I do get what you mean about clarity, though. What do you think?
Since this is related to config clarity, I'd also love y'all's input on this @fstavela @RonnyPfannschmidt @roliveri @FabriciaDinizRH

File: app/config.py (hunk 2 above)
Comment (kruai) [reply app/config.py:111]: Correct. For host payloads that don't have an `org_id`, we want to fetch it by default. We want that to go into effect as soon as possible, and then we can remove it once we start treating `org_id` like `account`.

File: app/instrumentation.py (hunk 3 above)
Comment (kruai) [reply app/instrumentation.py:190]: You're right, that was a mistake on my end. It should be a `Logger` instead of a `str`.

File: lib/middleware.py (hunk 5 above)
Comment (kruai) [reply lib/middleware.py:118]: The purpose of the POST is to get the `org_id` from the 3Scale org_id translation endpoint. That 3scale endpoint accepts an array of account numbers, and responds with a dict/map of the associated org_ids. 
From the BOP endpoint: `Dictionary of Organizations -> EBS Account number. Keys for non-existent mapping will be ommitted from the response`. So in that case, it will be None, and the host's `org_id` will remain None (unchanged from the input)

File: tests/test_mq_service.py [hunk 6]
```diff
@@ -1495,3 +1496,55 @@ def test_add_host_with_canonical_facts_MAC_address_valid_formats(mq_create_or_up
...
     assert created_host.mac_addresses == host_from_db.canonical_facts["mac_addresses"]
+
+
+def test_add_host_missing_org_id(mocker, mq_create_or_update_host, enable_org_id_translation):
+    """
+    Tests adding a host that's missing org_id.
+    Enables org_id translation, but patches the translator's response.
+    """
+    # Mock the post() functionality
+    session_post_mock = mocker.patch("lib.middleware.Session.post")
```
[the remark this answers is not in the document]
Comment (kruai) [reply tests/test_mq_service.py:1507]: I patched `lib.middleware.Session.post` and `lib.middleware.Session.mount` because we don't actually want this unit test to mount and POST a request to 3scale. Instead, on Line 1515, I tell it to mock the POST response using the format described in the response described in [the endpoint's swagger doc](https://backoffice-proxy.apps.ext.spoke.preprod.us-west-2.aws.paas.redhat.com/docs/#/default/accountMapFromOrgIDs).

So basically, this test is verifying that if org_id translation is enabled and the BOP endpoint responds in the expected format, `translate_account_to_org_id` will get the `org_id` from the response and set it on the Host.

File: Makefile (hunk 1 above)
Comment (fstavela) [reply Makefile:18]: Since we already use "BYPASS" vars for RBAC, I think that we should stay with `BYPASS_TENANT_TRANSLATION`, just to keep it consistent.

**Comment by kruai** [comment]
> @kruai Is there a documentation for the tenant translator somewhere?

The 3scale endpoint is described in Lindani's comment on this Jira: https://issues.redhat.com/browse/RHCLOUD-17654
That endpoint proxies to the BOP endpoint described here: https://backoffice-proxy.apps.ext.spoke.preprod.us-west-2.aws.paas.redhat.com/docs/#/default/accountMapFromOrgIDs

**Review by thearifismail — APPROVED** [review]

**Comment by kruai** [comment]
@fstavela I fixed the issue with the missing tenant-translator URL, and made sure we bypass that translation in local dev environments (since it needs to talk to 3scale internally)

**Comment by fstavela** [comment]
> @kruai The issue with the tenant-translator URL is fixed indeed. Now I see this error in the logs:
> 
> ```
> [2022-06-06 12:23:40,015] [1] [140498395508992] [inventory.app.queue.queue] [INFO] Attempting to add host
> [2022-06-06 12:23:40,032] [1] [140498395508992] [inventory.app.queue.queue] [ERROR] Error while adding host
> Traceback (most recent call last):
>   File "/opt/app-root/src/app/queue/queue.py", line 235, in add_host
>     output_host, host_id, insights_id, add_result = host_repository.add_host(
>   File "/opt/app-root/src/lib/host_repository.py", line 61, in add_host
>     input_host.org_id = translate_account_to_org_id(input_host.account)
>   File "/opt/app-root/src/lib/middleware.py", line 130, in translate_account_to_org_id
>     resp_data = translator_response.json()
>   File "/opt/app-root/lib64/python3.8/site-packages/requests/models.py", line 900, in json
>     return complexjson.loads(self.text, **kwargs)
>   File "/usr/lib64/python3.8/json/__init__.py", line 357, in loads
>     return _default_decoder.decode(s)
>   File "/usr/lib64/python3.8/json/decoder.py", line 337, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/usr/lib64/python3.8/json/decoder.py", line 355, in raw_decode
>     raise JSONDecodeError("Expecting value", s, err.value) from None
> json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> [2022-06-06 12:23:40,033] [1] [140498395508992] [inventory.app.queue.queue] [ERROR] Value error while adding or updating host: Expecting value: line 1 column 1 (char 0)
> [2022-06-06 12:23:40,033] [1] [140498395508992] [inventory.app.queue.queue] [ERROR] Unable to process message
> Traceback (most recent call last):
>   File "/opt/app-root/src/app/queue/queue.py", line 298, in event_loop
>     handler(message.value, event_producer)
>   File "<decorator-gen-6>", line 2, in handle_message
>   File "/opt/app-root/lib64/python3.8/site-packages/prometheus_client/context_managers.py", line 66, in wrapped
>     return func(*args, **kwargs)
>   File "/opt/app-root/src/app/queue/queue.py", line 272, in handle_message
>     output_host, host_id, insights_id, operation_result = message_operation(host, platform_metadata)
>   File "/opt/app-root/src/app/queue/queue.py", line 235, in add_host
>     output_host, host_id, insights_id, add_result = host_repository.add_host(
>   File "/opt/app-root/src/lib/host_repository.py", line 61, in add_host
>     input_host.org_id = translate_account_to_org_id(input_host.account)
>   File "/opt/app-root/src/lib/middleware.py", line 130, in translate_account_to_org_id
>     resp_data = translator_response.json()
>   File "/opt/app-root/lib64/python3.8/site-packages/requests/models.py", line 900, in json
>     return complexjson.loads(self.text, **kwargs)
>   File "/usr/lib64/python3.8/json/__init__.py", line 357, in loads
>     return _default_decoder.decode(s)
>   File "/usr/lib64/python3.8/json/decoder.py", line 337, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/usr/lib64/python3.8/json/decoder.py", line 355, in raw_decode
>     raise JSONDecodeError("Expecting value", s, err.value) from None
> json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> ```
> 
> I'm investigeting if it's an environment issue or application bug.

I tried to do `curl -X POST http://apicast.3scale-dev.svc.cluster.local:8892/internal/orgIds -d '["123456"]'` manually and I got error 500 with the message: `Invalid x-rh-insights-env stage specified.`, so I think it is an environment issue. I've already asked Lindani if he knows how to fix this: https://ansible.slack.com/archives/C022YV4E0NA/p1654520260195239

File: lib/middleware.py [hunk 7]
```diff
@@ -94,3 +101,32 @@ def modified_func(*args, **kwargs):
...
+
+def translate_account_to_org_id(account: str) -> str:
+    # If translation is bypassed, set the org_id to None
+    if inventory_config().bypass_tenant_translation:
+        return None
+
+    request_session = Session()
+    retry_config = Retry(total=inventory_config().rbac_retries, backoff_factor=1, status_forcelist=RETRY_STATUSES)
+    request_session.mount(tenant_translator_url(), HTTPAdapter(max_retries=retry_config))
+    body = dumps({"body": "[{" + account + "}]"})
```
Comment (fstavela) [inline lib/middleware.py:114]: @kruai Are you sure this is the correct data format for tenant translator? According to the examples in Jira and BOP documentation the endpoint only accepts a list of accounts. That list could be sent as `json` parameter of `Session.post` method.

Comment (kruai) [reply lib/middleware.py]: Oh you're right! I accidentally put some braces in there. I've updated it.

File: deployment.yaml [hunk 8]
```diff
@@ -716,9 +728,12 @@ parameters:
 - description: SSL validation mode for the DB
   name: INVENTORY_DB_SSL_MODE
   value: prefer
-- description: activate RBAC middleware
+- description: disable RBAC middleware
```
Comment (thearifismail) [inline deployment.yaml:731]: Should it be "enable" RBAC middleware because that's what the default value of "false" does?

Comment (kruai) [reply deployment.yaml:731]: That doesn't make sense to me... If `BYPASS_RBAC == True`, I'm disabling the RBAC middleware. I don't think the default value should influence the description of the variable

Comment (thearifismail) [reply deployment.yaml:731]: yes agree with default value should not make the difference but find the description confusing enough to comment on it.

**Review by thearifismail — COMMENTED** [review]
LGTM one last comment in `test_add_host_missing_org_id_translation`

File: tests/test_mq_service.py [hunk 9]
```diff
@@ -1495,3 +1496,93 @@ def test_add_host_with_canonical_facts_MAC_address_valid_formats(mq_create_or_up
     host_from_db = db_get_host(created_host.id)
 
     assert created_host.mac_addresses == host_from_db.canonical_facts["mac_addresses"]
+
+
+def test_add_host_missing_org_id_translation(mocker, mq_create_or_update_host, db_get_host, enable_org_id_translation):
+    """
+    Enables org_id translation, but patches the translator's response.
+    Tests adding a host that's missing org_id.
```
[the remark this answers is not in the document]
Comment (kruai) [reply tests/test_mq_service.py:1504]: @thearifismail I'm unclear on what you're asking for... are you requesting the comment be updated, or the test name?

[the remark this answers is not in the document]
Comment (thearifismail) [reply tests/test_mq_service.py:1504]: The tests does not do what the comment says.  As far the tests itself is concerned, it is not needed because the functionality is covered by `test_add_org_id_to_existing_host`

**Review by thearifismail — APPROVED** [review]

File: tests/test_mq_service.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kruai) [reply tests/test_mq_service.py:1504]: We resolved this by discussing in-person, but here's a summary or posterity:
`test_add_org_id_to_existing_host` has OrgID translation disabled, while `test_add_host_missing_org_id_translation` has it enabled and is specifically testing that middleware functionality.
